File: forwarder/app/src/main/java/com/codey/forwarder/ForwarderNotificationListenerService.java
package com.codey.forwarder;

import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;

import org.json.JSONObject;

public class ForwarderNotificationListenerService extends NotificationListenerService {
    @Override
    public void onListenerConnected() {
        super.onListenerConnected();
        ForwarderConfig.saveStatus(this, "Notification listener connected.");
    }

    @Override
    public void onNotificationPosted(StatusBarNotification sbn) {
        ForwarderSettings settings = ForwarderConfig.readSettings(this);
        if (!settings.forwardEnabled) {
            return;
        }
        if (!ForwarderNotifier.shouldForwardPackage(sbn.getPackageName(), settings.forwardBusiness)) {
            return;
        }

        ForwardedNotification notification = ForwarderNotifier.buildForwardedNotification(sbn);
        new Thread(() -> forward(settings, notification)).start();
    }

    private void forward(ForwarderSettings settings, ForwardedNotification notification) {
        try {
            JSONObject payload = ForwarderNotifier.buildForwarderPayload(settings, notification);
            ForwarderNotifier.HttpResult result =
                ForwarderNotifier.postNotificationPayload(settings.webhookUrl, payload);
            String message = result.statusCode >= 200 && result.statusCode <= 299
                ? "Forwarded WhatsApp notification: HTTP " + result.statusCode
                : "Forward failed: HTTP " + result.statusCode + " " + take(result.responseBody, 160);
            ForwarderConfig.saveStatus(
                this,
                message,
                notification.title,
                notification.body,
                System.currentTimeMillis()
            );
        } catch (Exception error) {
            ForwarderConfig.saveStatus(
                this,
                "Forward failed: " + (error.getMessage() != null
                    ? error.getMessage()
                    : error.getClass().getSimpleName()),
                notification.title,
                notification.body,
                System.currentTimeMillis()
            );
        }
    }

    private static String take(String value, int length) {
        if (value == null) {
            return "";
        }
        return value.length() <= length ? value : value.substring(0, length);
    }
}

File: forwarder/app/src/main/java/com/codey/forwarder/ForwarderNotificationListenerService.kt
package com.codey.app

import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification

class ForwarderNotificationListenerService : NotificationListenerService() {
    override fun onListenerConnected() {
        super.onListenerConnected()
        ForwarderConfig.saveStatus(this, "Notification listener connected.")
    }

    override fun onNotificationPosted(sbn: StatusBarNotification) {
        val settings = ForwarderConfig.readSettings(this)
        if (!settings.forwardEnabled) {
            return
        }
        if (!ForwarderNotifier.shouldForwardPackage(sbn.packageName, settings.forwardBusiness)) {
            return
        }

        val notification = ForwarderNotifier.buildForwardedNotification(sbn)
        Thread { forward(settings, notification) }.start()
    }

    private fun forward(settings: ForwarderSettings, notification: ForwardedNotification) {
        try {
            val payload = ForwarderNotifier.buildForwarderPayload(settings, notification)
            val result = ForwarderNotifier.postNotificationPayload(
                ForwarderNotifier.resolveNotificationWebhookUrl(settings),
                payload,
                settings.deviceToken
            )
            val message = if (result.statusCode in 200..299) {
                "Forwarded WhatsApp notification: HTTP ${result.statusCode}"
            } else {
                "Forward failed: HTTP ${result.statusCode} ${result.responseBody.takeAtMost(160)}"
            }
            ForwarderConfig.saveStatus(
                this,
                message,
                notification.title,
                notification.body,
                System.currentTimeMillis()
            )
        } catch (error: Exception) {
            ForwarderConfig.saveStatus(
                this,
                "Forward failed: ${error.safeMessage()}",
                notification.title,
                notification.body,
                System.currentTimeMillis()
            )
        }
    }
}
